src/v/compression/internal/lz4_frame_compressor.cc

```diff
@@ -121,112 +170,116 @@ compute_frame_uncompressed_size(size_t frame_size, size_t original) {
     return frame_size;
 }
 
-static iobuf do_uncompressed(const char* src, const size_t src_size) {
-    auto ctx_ptr = make_decompression_context();
-    LZ4F_decompressionContext_t ctx = ctx_ptr.get();
-    LZ4F_frameInfo_t fi;
-    size_t in_sz = src_size;
-    LZ4F_errorCode_t code = LZ4F_getFrameInfo(ctx, &fi, src, &in_sz);
-    check_lz4_error("lz4f_getframeinfo error: {}", code);
-    size_t this_chunk_size = compute_frame_uncompressed_size(
-      fi.contentSize, src_size);
-
-    size_t max_chunk = 128_KiB;
-    this_chunk_size = std::min(this_chunk_size, max_chunk);
-
-    // We will decompress into temporary_buffers, and each time
-    // we spill past estimated_output_size, push the
-    ss::temporary_buffer<char> obuf(this_chunk_size);
-    char* out = obuf.get_write();
-
-    // Offset in `src` buffer
-    size_t src_offset = in_sz;
+iobuf lz4_frame_compressor::uncompress(iobuf const& input) {
+    size_t src_size = input.size_bytes();
 
-    // Offset in `out` buffer
-    size_t output_this_chunk{0};
+    const size_t max_chunk = details::io_allocation_size::max_chunk_size;
 
     // Compose fragmented result from a series of `obuf` temporary buffers
     iobuf ret;
 
-    while (src_offset < src_size) {
+    // Read interators/offsets
+    auto frag_i = input.begin();
+    size_t read_this_chunk{0};
+    size_t read_total{0};
+
+    auto ctx_ptr = make_decompression_context();
+    LZ4F_decompressionContext_t ctx = ctx_ptr.get();
+
+    // Prior to main loop, optionally consume header to learn total size
+    LZ4F_errorCode_t code = 0;
+    size_t decompressed_size = 0;
+    auto& first_chunk = *(input.begin());
```

Comment:
would this be undefined behavior if input were an empty iobuf?

Reply:
Updated this to explicitly check for the empty case